firebase-firestore/src/main/java/com/google/firebase/firestore/auth/FirebaseAuthCredentialsProvider.java

```diff
@@ -60,30 +64,47 @@
   private boolean forceRefresh;
 
   /** Creates a new FirebaseAuthCredentialsProvider. */
-  public FirebaseAuthCredentialsProvider(InternalAuthProvider authProvider) {
-    this.authProvider = authProvider;
+  public FirebaseAuthCredentialsProvider(Deferred<InternalAuthProvider> deferredAuthProvider) {
     this.idTokenListener =
         token -> {
           synchronized (this) {
-            currentUser = getUser();
+            currentUser = getUser(internalAuthProviderProviderRef.get());
             tokenCounter++;
 
             if (changeListener != null) {
               changeListener.onValue(currentUser);
             }
           }
         };
-    currentUser = getUser();
+
+    internalAuthProviderProviderRef = new AtomicReference<>(() -> null);
+    currentUser = getUser(internalAuthProviderProviderRef.get());
     tokenCounter = 0;
 
-    authProvider.addIdTokenListener(idTokenListener);
+    deferredAuthProvider.whenAvailable(
+        provider -> {
+          synchronized (this) {
+            internalAuthProviderProviderRef.set(provider);
+            currentUser = getUser(internalAuthProviderProviderRef.get());
+            if (changeListener != null) {
+              changeListener.onValue(currentUser);
+            }
+          }
+          provider.get().addIdTokenListener(idTokenListener);
+        });
   }
 
   @Override
   public synchronized Task<String> getToken() {
     boolean doForceRefresh = forceRefresh;
     forceRefresh = false;
-    Task<GetTokenResult> res = authProvider.getAccessToken(doForceRefresh);
+
+    InternalAuthProvider internalAuthProvider = internalAuthProviderProviderRef.get().get();
+    if (internalAuthProvider == null) {
+      return Tasks.forException(new FirebaseApiNotAvailableException("auth is not available"));
```

Comment:
Are you sure you want to throw in this case? Shouldn't this case be treated the same as it is in the EmptyCredentialsProvider and return `Task.forResult(null)`?
https://github.com/firebase/firebase-android-sdk/blob/18d574fe59119c4422a3acf4a566c20bd187e0fb/firebase-firestore/src/main/java/com/google/firebase/firestore/auth/EmptyCredentialsProvider.java#L25-L29

Additionally, is `EmptyCredentialsProvider` still needed or can it be removed?

Reply:
What are your thoughts on this in light of the code in `FirestoreCallCredentials` that I linked to?

Reply:
Your answer wasn't quite what I predicted. You have changed my mind. I was expecting the consumer to be a bit simpler than it actually is. Please disregard.

Reply:
Ok. I've left the `FirebaseApiNotAvailableException` in there.